core/drivers/crypto/hisilicon/sec_cipher.c

```diff
@@ -0,0 +1,776 @@
+		break;
+	case C_ALG_3DES:
+		ret = sec_cipher_3des_get_c_key_len(key_len, &c_key_len);
+		break;
+	case C_ALG_AES:
+		ret = sec_cipher_aes_get_c_key_len(key_len, c_ctx->mode,
+						   &c_key_len);
+		break;
+	case C_ALG_SM4:
+		ret = sec_cipher_sm4_get_c_key_len(key_len, c_ctx->mode,
+						   &c_key_len);
+		break;
+	default:
+		EMSG("Invalid cipher type %#"PRIx8, c_ctx->alg);
+		ret = TEE_ERROR_NOT_IMPLEMENTED;
+		break;
+	}
+
+	if (ret)
+		return ret;
+
+	c_ctx->key_dma = virt_to_phys(c_ctx->key);
+	if (!c_ctx->key_dma) {
+		EMSG("c_key_dma is NULL");
+		return TEE_ERROR_STORAGE_NO_SPACE;
+	}
+
+	c_ctx->key_len = key_len;
+	c_ctx->c_key_len = c_key_len;
+
+	memcpy(c_ctx->key, key1, key1_len);
+	memcpy(c_ctx->key + key1_len, key2, key2_len);
+
+	return ret;
+}
+
+static void ctr_iv_inc(uint8_t *counter, uint32_t shift_len)
+{
+	uint32_t n = CTR_128BIT_COUNTER;
+	uint8_t *counter1 = counter;
+	uint32_t c = shift_len;
+
+	/* increment counter (128-bit int) by c */
+	do {
+		--n;
+		c += counter1[n];
+		counter1[n] = (uint8_t)c;
+		c >>= BYTE_BITS;
+	} while (n);
+}
+
+static void xts_multi_galois(unsigned char *data)
+{
+	int i = 0;
+	uint8_t t = 0;
+	uint8_t tt = 0;
+
+	for (; i < AES_SM4_IV_SIZE; i++) {
+		tt = data[i] >> LEFT_MOST_BIT;
+		data[i] = ((data[i] << 1) | t) & 0xFF;
+		t = tt;
+	}
+	if (tt)
+		data[0] ^= 0x87;
+}
+
+/*
+ * When the IV is delivered by segment,
+ * the AES/SM4-ECB is used to update the IV to be used next time.
+ */
+static TEE_Result xts_iv_update(struct sec_cipher_ctx *c_ctx)
+{
+	size_t xts_key_len = c_ctx->key_len >> 1;
+	struct sec_cipher_ctx ecb_ctx = {0};
+	TEE_Result ret = TEE_SUCCESS;
+	size_t i = 0;
+
+	ecb_ctx.alg = c_ctx->alg;
+	ecb_ctx.mode = C_MODE_ECB;
+	ret = sec_cipher_set_key(&ecb_ctx, c_ctx->key + xts_key_len,
+				 xts_key_len, NULL, 0);
+	if (ret)
+		return ret;
+
+	ecb_ctx.encrypt = true;
+	ecb_ctx.in = c_ctx->iv;
+	ecb_ctx.out = c_ctx->iv;
+	ecb_ctx.in_dma = c_ctx->iv_dma;
+	ecb_ctx.out_dma = c_ctx->iv_dma;
+	ecb_ctx.len = c_ctx->iv_len;
+
+	ret = sec_do_cipher_task(c_ctx->qp, &ecb_ctx);
+	if (ret != 0) {
+		EMSG("Fail to encrypt xts iv, ret=%#"PRIx32, ret);
+		return ret;
+	}
+
+	for (i = 0; i < DIV_ROUND_UP(c_ctx->len, AES_SM4_BLOCK_SIZE); i++)
+		xts_multi_galois(c_ctx->iv);
+
+	ecb_ctx.encrypt = false;
+	ret = sec_do_cipher_task(c_ctx->qp, &ecb_ctx);
+	if (ret != 0)
+		EMSG("Fail to decrypt xts iv, ret=%#"PRIx32, ret);
+
+	return ret;
+}
+
+static TEE_Result sec_update_iv(struct sec_cipher_ctx *c_ctx)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	size_t offset = 0;
+
+	switch (c_ctx->mode) {
+	case C_MODE_CBC:
+		offset = c_ctx->len - c_ctx->iv_len;
+		if (c_ctx->encrypt && c_ctx->len >= c_ctx->iv_len)
+			memcpy(c_ctx->iv, c_ctx->out + offset, c_ctx->iv_len);
+		if (!c_ctx->encrypt && c_ctx->len >= c_ctx->iv_len)
+			memcpy(c_ctx->iv, c_ctx->in + offset, c_ctx->iv_len);
+		break;
+	case C_MODE_CTR:
+		ctr_iv_inc(c_ctx->iv, c_ctx->len >> CTR_MODE_LEN_SHIFT);
```

Comment:
Please add a comment explaining why `c_ctx->len` should be shifted with this amount.
What is the relationship between `c_ctx->len` and the amount you're increasing by?

Reply:
I can't make sense of that, but it seems you'd like to increase the IV counter with the number of processed blocks.
If so, please explain that with a comment in the code.

Reply:
That's what I meant. I'll add a comment.